### PR TITLE
Adding ability to update docProps

### PR DESCRIPTION
**Pull request by sbeih**
- fixed `createdBy` , `lastModifiedBy` , `created` and `modified` doc props.
- Added ability to set custom docProps.

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+export interface QueryInfo {
```
Comment (liorZilberstein) [inline src/types.ts:1]: Add a Copyright statement

File: src/workbookManager.ts [hunk 2]
```diff
@@ -79,7 +154,10 @@ export class WorkbookManager {
             if (
                 properties.getAttribute("command") == "SELECT * FROM [Query1]"
             ) {
-                properties.parentElement?.setAttribute("refreshOnLoad", "1");
+                properties.parentElement?.setAttribute(
```
Comment (liorZilberstein) [inline src/workbookManager.ts:157]: Do we want to always set this property? why did you change it from only when needed to always?

File: src/workbookManager.ts [hunk 3]
```diff
@@ -34,31 +36,97 @@ export class WorkbookManager {
...
+        await this.updateSingleQueryRefreshOnOpen(zip, query.refreshOnOpen);
+        await this.updateDocProps(zip, docProps);
+
+        return await zip.generateAsync({
+            type: "blob",
+            mimeType: "application/xlsx",
+        });
+    }
+
+    private async updatePowerQueryDocument(zip: JSZip, queryMashup: string) {
```
[the remark this answers is not in the document]
Comment (sbeih) [reply src/workbookManager.ts:57]: I don't see how docProps can be a part of `PowerQueryDocument`, isn't it a workbook-level data?

File: src/workbookManager.ts [hunk 4]
```diff
@@ -34,31 +36,97 @@ export class WorkbookManager {
...
+        if (!docProps.createdBy) docProps.createdBy = defaultDocProps.createdBy;
+        if (!docProps.lastModifiedBy)
+            docProps.lastModifiedBy = defaultDocProps.lastModifiedBy;
+
+            //set auto updated elements
+        const { doc, properties } = await documentUtils.getDocPropsProperties(
+            zip
+        );
+
+        const docPropsAutoUpdatedElementsArr = Object.keys(
```
[the remark this answers is not in the document]
Comment (sbeih) [reply src/workbookManager.ts:80]: It would be executed only once, since this api just generates a new workbook.
 I updated the dates in the new default template PR, but I think it would still be better if every time we generate a workbook we use current time of generation.

File: src/workbookManager.ts (hunk 2 above)
Comment (sbeih) [reply src/workbookManager.ts]: @ron-b this change was in your code, I'm not actually sure what did you solve here, can you please check this?

File: package-lock.json [hunk 5]
```diff
@@ -1,8 +1,9482 @@
 {
   "name": "@microsoft/connected-workbooks",
   "version": "1.0.0",
-  "lockfileVersion": 1,
```
Comment (liorZilberstein) [inline package-lock.json:4]: why is this file part of the change?

**Review by liorZilberstein — APPROVED** [review]

File: src/utils/documentUtils.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+// Licensed under the MIT license.
+
+import JSZip from "jszip";
+import { docPropsCoreXmlPath, docPropsRootElement } from "../constants";
+
+export const createOrUpdateProperty = (
+    doc: Document,
+    parent: Element,
+    property: string,
+    value: string | null | undefined
```
Comment (ron-b) [inline src/utils/documentUtils.ts:11]: small pattern mix - once we do `x?: string | null` and another time `x: string | null | undefined`

File: src/utils/documentUtils.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+): void => {
+    if (!value) {
```
Comment (ron-b) [inline src/utils/documentUtils.ts:13]: this `if(!value)` check catches both the case for null and undefined.
My intent was to have null clear out the value if it exists, and undefined to be no-op for a specific property.

File: src/utils/documentUtils.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        .file(docPropsCoreXmlPath)
+        ?.async("text");
+    if (docPropsCoreXmlString === undefined) {
+        throw new Error("DocProps core.xml was not found in template");
+    }
+
+    const parser = new DOMParser();
+    const doc = parser.parseFromString(docPropsCoreXmlString, "text/xml");
+
+    const properties = doc.getElementsByTagName(docPropsRootElement).item(0);
```
Comment (ron-b) [inline src/utils/documentUtils.ts:44]: `doc` isn't guaranteed, right? also - `item(0)`

Comment (ron-b) [reply src/utils/documentUtils.ts:44]: Does it make sense to have it as `doc?.getElementsByTagName(docPropsRootElement)?.item(0);`

File: src/workbookManager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ron-b) [reply src/workbookManager.ts:55]: It is workbook level data, I commented towards refactoring, which looks good now.

File: src/workbookManager.ts [hunk 9]
```diff
@@ -34,31 +34,92 @@ export class WorkbookManager {
...
+            docPropsAutoUpdatedElements
+        ) as Array<keyof typeof docPropsAutoUpdatedElements>;
+
+        const nowTime = new Date().toISOString();
+
+        docPropsAutoUpdatedElementsArr.forEach((tag) => {
+            if (
+                properties.getElementsByTagName(
+                    docPropsAutoUpdatedElements[tag]
+                ).length !== 1
```
Comment (ron-b) [inline src/workbookManager.ts:81]: WRT docPropsAutoUpdatedElements - when one is missing does Excel fail to open the workbook? if not - I don't think its invalid.
What happens if there are 2 or more of the same?

Questions are mainly to see we align to document validity Office currently has.

File: src/workbookManager.ts [hunk 10]
```diff
@@ -67,26 +128,30 @@ export class WorkbookManager {
...
-                const newConn = serializer.serializeToString(connectionsDoc);
-                zip.file(connectionsXmlPath, newConn);
-                break;
-            }
+
+        const connectionsPropertiesArr = [...connectionsProperties];
+        const queryProp = connectionsPropertiesArr.find((prop) => {
+            prop.getAttribute("command") == "SELECT * FROM [Query1]";
+        });
+        if (queryProp) {
```
Comment (ron-b) [inline src/workbookManager.ts:144]: if queryProp isn't found - error

File: src/workbookManager.ts [hunk 11]
```diff
@@ -67,26 +128,30 @@ export class WorkbookManager {
         }
         const parser: DOMParser = new DOMParser();
         const serializer = new XMLSerializer();
-
+        const refreshOnLoadValue = refreshOnOpen ? "1" : "0";
         const connectionsDoc: Document = parser.parseFromString(
             connectionsXmlString,
             "text/xml"
         );
         let connectionId = "-1";
```
Comment (ron-b) [inline src/workbookManager.ts:136]: might be a cleaner solution than what I've introduced originally - leave it undefined and if you reach current check - that means we're good to continue (since we throw otherwise)

File: src/workbookManager.ts [hunk 12]
```diff
@@ -67,26 +128,30 @@ export class WorkbookManager {
...
-                zip.file(connectionsXmlPath, newConn);
-                break;
-            }
+
+        const connectionsPropertiesArr = [...connectionsProperties];
+        const queryProp = connectionsPropertiesArr.find((prop) => {
+            prop.getAttribute("command") == "SELECT * FROM [Query1]";
+        });
+        if (queryProp) {
+            queryProp.parentElement?.setAttribute(
```
Comment (ron-b) [inline src/workbookManager.ts:145]: does this need to be `queryProp.parentElement!.`?

File: src/workbookManager.ts (hunk 2 above)
Comment (ron-b) [reply src/workbookManager.ts]: @sbeih I believe this was a bug fix to address custom templates

**Comment by ron-b** [comment]
I believe tests to the docProps editing layer can be added in scope of this PR or in a following one, scoped to testing out just that xml piece without having the need to validate the Excel file structure.

File: package-lock.json (hunk 5 above)
Comment (sbeih) [reply package-lock.json:4]: New `package-lock.json` format in npmV7, you can check the changes [here](https://blog.npmjs.org/post/621733939456933888/npm-v7-series-why-keep-package-lockjson.html)

File: src/utils/documentUtils.ts (hunk 8 above)
Comment (sbeih) [reply src/utils/documentUtils.ts:44]: `doc` is a required argument to the function, so it must be a Document object.

File: src/workbookManager.ts (hunk 9 above)
Comment (sbeih) [reply src/workbookManager.ts:81]: They already exists in the default template, we just update them here (dates), 
if there are 2 or more of the same, value get's overrided by the later

Comment (ron-b) [reply src/workbookManager.ts:81]: Since we already have some support for custom templates, input workbook props can vary, best validate how Excel behaves today with these kind of cases and align.

File: src/utils/documentUtils.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        throw new Error(
+            `Invalid DocProps core.xml, multiple ${property} elements`
+        );
+    } else if (elements?.length > 0) {
+        elements[0]!.textContent = value;
+        elements[0]!.innerHTML = value!;
+    }
+};
+
+export const getDocPropsProperties = async (
```
Comment (ron-b) [inline src/utils/documentUtils.ts:33]: is this intentional? (do we need both export and export default?)

File: src/workbookManager.ts [hunk 14]
```diff
@@ -1,30 +1,30 @@
...
-    refreshOnOpen: boolean;
-    constructor(queryMashup: string, refreshOnOpen: boolean) {
-        this.queryMashup = queryMashup;
-        this.refreshOnOpen = refreshOnOpen;
-    }
-}
 export class WorkbookManager {
     private mashupHandler: MashupHandler = new MashupHandler();
 
-    async generateSingleQueryWorkbook(
```
Comment (ron-b) [inline src/workbookManager.ts:25]: Please keep a wrapper with the original name for now, for back compat

File: src/workbookManager.ts (hunk 9 above)
Comment (ron-b) [reply src/workbookManager.ts:88]: To be clear - lets verify these fields are mandatory before we add strict validation for them in templates.

**Review by ron-b — APPROVED** [review]
